File: 2812. Find the Safest Path in a Grid.py
class Solution:
    def maximumSafenessFactor(self, grid: List[List[int]]) -> int:
        n = len(grid)
        if n == 1: return 0
        if grid[0][0] == 1 or grid[n-1][n-1] == 1: return 0
        min_distances = self.find_min(grid)
        left, right = 0, n*n
        answer = 0
        visited = set()
        while left <= right:
            mid = (left + right) // 2
            if self.can_do(min_distances, mid):
                answer = mid
                left = mid + 1
            else:
                right = mid - 1
        return answer
    def can_do(self, min_distances, limit):
        n = len(min_distances)
        if min_distances[0][0] < limit:
            return False
        if min_distances[n-1][n-1] < limit:
            return False
        visited = set()
        queue = deque([(0, 0)])
        while queue:
            x, y = queue.popleft()
            if x == n-1 and y == n-1:
                return True
            if (x, y) in visited:
                continue
            visited.add((x, y))
            for dx, dy in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
                nx, ny = x + dx, y + dy
                if 0 <= nx < n and 0 <= ny < n and (nx, ny) not in visited and min_distances[nx][ny] >= limit:
                    queue.append((nx, ny))
        return False
    def find_min(self, grid):
        n = len(grid)
        queue = deque()
        for i in range(n):
            for j in range(n):
                if grid[i][j] == 1:
                    queue.append((i, j))
        min_distances = [[0]*n for _ in range(n)]
        distance = 0
        visited = set()
        while queue:
            size = len(queue)
            while size > 0:
                x, y = queue.popleft()
                size -= 1
                if (x, y) in visited: continue
                visited.add((x, y))
                min_distances[x][y] = distance
                for dx, dy in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
                    nx, ny = x + dx, y + dy
                    if 0 <= nx < n and 0 <= ny < n and (nx, ny) not in visited:
                        queue.append((nx, ny))
            distance += 1
        return min_distances

File: 2812. Find the Safest Path in a Grid.java
class Solution {
    private static final int[][] directions = {{0,1},{0,-1},{1,0},{-1,0}};
    public int maximumSafenessFactor(List<List<Integer>> grid) {
        int n = grid.size();
        if (n == 1) return 0;
        if (grid.get(n-1).get(n-1) == 1 || grid.get(0).get(0) == 1) return 0;
        int[][] minDistances = findMin(grid);
        int left = 0;
        int right = grid.size();
        int answer = Integer.MIN_VALUE;
        boolean[][] visited = new boolean[grid.size()][grid.size()];
        while (left <= right) {
            int mid = (left + right) / 2;
            if (canDo(minDistances, mid)) {
                answer = mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return answer;
    }
    private int[][] findMin(List<List<Integer>> grid) {
        int n = grid.size();
        Deque<int[]> bfsQ = new ArrayDeque<>();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (grid.get(i).get(j) == 1) {
                    bfsQ.addLast(new int[] {i, j});
                }
            }
        }
        int[][] minDistance = new int[n][n];
        boolean[][] visited = new boolean[n][n];
        int distance = 0;
        while (!bfsQ.isEmpty()) {
            int size = bfsQ.size();
            while (size-- > 0) {
                int[] current = bfsQ.pollFirst();
                if (visited[current[0]][current[1]]) continue;
                visited[current[0]][current[1]] = true;
                minDistance[current[0]][current[1]] = distance;
                for (int[] d: directions) {
                    int nextX = current[0] + d[0];
                    int nextY = current[1] + d[1];
                    if (nextX < 0 || nextX >= grid.size()) continue;
                    if (nextY < 0 || nextY >= grid.size()) continue;
                    if (visited[nextX][nextY]) continue;
                    bfsQ.addLast(new int[] {nextX, nextY});
                }
            }
            distance++;
        }
        return minDistance;
    }
    private boolean canDo(int[][] minDistances, int limit) {
        int n = minDistances.length;
        if (minDistances[0][0] < limit || minDistances[n-1][n-1] < limit) return false;
        Deque<int[]> bfsQ = new ArrayDeque<>();
        bfsQ.addLast(new int[] {0, 0});
        boolean[][] visited = new boolean[n][n];
        while (!bfsQ.isEmpty()) {
            int[] current = bfsQ.pollFirst();
            if (current[0] == n-1 && current[1] == n-1) return true;
            if (visited[current[0]][current[1]]) continue;
            visited[current[0]][current[1]] = true;
            for (int[] d : directions) {
                int nextX = current[0] + d[0];
                int nextY = current[1] + d[1];
                if (nextX < 0 || nextX >= n) continue;
                if (nextY < 0 || nextY >= n) continue;
                if (visited[nextX][nextY]) continue;
                if (minDistances[nextX][nextY] < limit) continue;
                bfsQ.addLast(new int[] {nextX, nextY});
            }
        }
        return false;
    }
}